### PR TITLE
Feature/hiera

### PR DESCRIPTION
**Pull request by benjijamorris**
## What does this PR do?
- add 2d and 3d Hiera Models
- refactor mae/hiera to encoder/decoder.py instead of model-specific files
- unify patchify code
- add tests
- updates to JEPA to accommodate refactoring

## Before submitting

- [x] Did you make sure **title is self-explanatory** and **the description concisely explains the PR**?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **test your PR locally** with `pytest` command?
- [x] Did you **run pre-commit hooks** with `pre-commit run -a` command?

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: cyto_dl/nn/vits/blocks/masked_unit_attention.py [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+
+        q, k, v = qkv[0], qkv[1], qkv[2]
+
+        if np.any(self.q_stride > 1):
+            # within a mask unit, tokens are spatially ordered
+            # perform spatial 2x2x2 max pooling over tokens
+            if self.spatial_dims == 3:
+                q = reduce(
+                    q,
+                    "b n h (n_patches_z q_stride_z n_patches_y q_stride_y n_patches_x q_stride_x) c ->b n h (n_patches_z n_patches_y n_patches_x) c",
```
[the remark this answers is not in the document]
Comment (ritvikvasan) [reply cyto_dl/nn/vits/blocks/masked_unit_attention.py:90]: c = head_dim

File: cyto_dl/nn/vits/blocks/masked_unit_attention.py [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """
+        x: batch x mask units x tokens x emb_dim
+        """
+        # Attention + Q Pooling
+        x_norm = self.norm1(x)
+        # change dimension and subsample within mask unit for skip connection
+        x = self.proj(x_norm)
+
+        x = x + self.drop_path(self.attn(x_norm))
```
Comment (ritvikvasan) [inline cyto_dl/nn/vits/blocks/masked_unit_attention.py:229]: Does dim_out = dim for skip connection with attention?

File: cyto_dl/nn/vits/blocks/masked_unit_attention.py [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+
+
+class HieraBlock(nn.Module):
+    def __init__(
+        self,
+        dim: int,
+        dim_out: int,
+        heads: int,
+        spatial_dims: int = 3,
+        mlp_ratio: float = 4.0,
```
Comment (ritvikvasan) [inline cyto_dl/nn/vits/blocks/masked_unit_attention.py:133]: what is mlp_ratio? add to docstring?

File: cyto_dl/nn/vits/blocks/patchify/patchify_hiera.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+            b=sequences.shape[0],
+            c=sequences.shape[-1],
+            mu=sequences.shape[1],
+            p=sequences.shape[2],
+        ),
+    )
+
+
+class PatchifyHiera(PatchifyBase):
+    """Class for converting images to a masked sequence of patches with positional embeddings."""
```
Comment (ritvikvasan) [inline cyto_dl/nn/vits/blocks/patchify/patchify_hiera.py:33]: to "mask units" instead of masked sequence? since that's what a regular patchify does?

File: cyto_dl/nn/vits/utils.py [hunk 5]
```diff
@@ -40,3 +47,8 @@ def get_positional_embedding(
             cls_token = torch.zeros(1, 1, emb_dim)
             pe = torch.cat([cls_token, pe], dim=0)
         return torch.nn.Parameter(pe, requires_grad=False)
+
+
+def validate_spatial_dims(spatial_dims, tuples):
```
Comment (ritvikvasan) [inline cyto_dl/nn/vits/utils.py:52]: I feel like the code might be clearer by not having this be a separate function and just calling these 2 lines in every class? I thought this function was doing a lot more based on the name (like some math to check that the spatial dimensions of each patch and mask is correct). What do you think?

**Review by ritvikvasan — DISMISSED** [review]
This is a lot!! Loved reading it...As always, minor comments

**Comment by ritvikvasan** [comment]
windows tests always seem to be failing. any ideas why?

File: configs/model/im2im/hiera.yaml [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+save_images_every_n_epochs: 1
+save_dir: ${paths.output_dir}
+
+x_key: ${source_col}
+
+backbone:
+  _target_: cyto_dl.nn.vits.mae.HieraMAE
+  spatial_dims: ${spatial_dims}
+  patch_size: 2 # patch_size* num_patches should be your patch shape
```
[the remark this answers is not in the document]
Comment (benjijamorris) [reply configs/model/im2im/hiera.yaml:11]: yes - the terminology is confusing here haha. "patch" = the small crop extracted from your original image, but "patch" is also the tokenized component of the image fed into the network. The patch size can be either an int (repeated for each spatial dim) or a list of size spatial_dims

File: configs/model/im2im/hiera.yaml [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    - repeat: 1
+      q_stride: 2
+      num_heads: 1
+    - repeat: 1
+      q_stride: 1
+      num_heads: 2
+    # self attention transformer - attention is done across all patches, irrespective of which mask unit they're in
+    - repeat: 2
+      num_heads: 4
+      self_attention: True
```
[the remark this answers is not in the document]
Comment (benjijamorris) [reply configs/model/im2im/hiera.yaml:27]: correct. 3 layers is small enough to test quickly. All of the models with unit tests are tiny by default in the configs and I have somewhere in the docs that you should increase the model size if you want good performance.

**Comment by benjijamorris** [comment]
the windows tests are just way slower for some reason... all the tests pass but I set a 70 minute time out so we don't rack up crazy costs.

File: cyto_dl/nn/vits/blocks/masked_unit_attention.py (hunk 2 above)
Comment (benjijamorris) [reply cyto_dl/nn/vits/blocks/masked_unit_attention.py:229]: Good question - each block specified in the architecture argument doubles the embedding dimension and halves the size of the mask unit. This doubling/pooling happens on the last repeat of the block, so dim_out=dim for all repeats except the last. I updated the docstring with an example.

File: cyto_dl/nn/vits/utils.py (hunk 5 above)
Comment (benjijamorris) [reply cyto_dl/nn/vits/utils.py:52]: I'd prefer renaming it to something clearer rather than repeating the code, maybe `match_tuple_to_spatial_dims`?

Comment (ritvikvasan) [reply cyto_dl/nn/vits/utils.py:52]: sounds good

File: cyto_dl/nn/vits/blocks/masked_unit_attention.py (hunk 2 above)
Comment (ritvikvasan) [reply cyto_dl/nn/vits/blocks/masked_unit_attention.py:226]: cool!

**Review by ritvikvasan — DISMISSED** [review]

**Review by ritvikvasan — APPROVED** [review]
